include/mbgl/style/expression/match.hpp

```diff
@@ -0,0 +1,277 @@
+        }
+    }
+    
+    bool isFeatureConstant() const override {
+        if (!input->isFeatureConstant()) { return false; }
+        if (!otherwise->isFeatureConstant()) { return false; }
+        for (const auto& pair : cases) {
+            if (!pair.second->isFeatureConstant()) { return false; }
+        }
+        return true;
+    }
+
+    bool isZoomConstant() const override {
+        if (!input->isZoomConstant()) { return false; }
+        if (!otherwise->isZoomConstant()) { return false; }
+        for (const auto& pair : cases) {
+            if (!pair.second->isZoomConstant()) { return false; }
+        }
+        return true;
+    }
+    
+    EvaluationResult evaluate(const EvaluationParameters& params) const override {
+        const auto& inputValue = evaluateInput(params);
+        if (!inputValue) {
+            return inputValue.error();
+        }
+        if (cases.find(*inputValue) == cases.end()) {
+            return otherwise->evaluate(params);
+        }
+        return cases.at(*inputValue)->evaluate(params);
+    }
+
+private:
+    Result<T> evaluateInput(const EvaluationParameters& params) const;
+
+    std::unique_ptr<TypedExpression> input;
+    std::unordered_map<T, std::shared_ptr<TypedExpression>> cases;
+    std::unique_ptr<TypedExpression> otherwise;
+};
+
+class UntypedMatch : public UntypedExpression {
+public:
+    using Cases = std::vector<std::pair<MatchKey, std::unique_ptr<UntypedExpression>>>;
+    
+    UntypedMatch(std::string key,
+                 std::unique_ptr<UntypedExpression> input_,
+                 Cases cases_,
+                 std::unique_ptr<UntypedExpression> otherwise_,
+                 const type::Type& inputType_) :
+        UntypedExpression(key),
+        input(std::move(input_)),
+        cases(std::move(cases_)),
+        otherwise(std::move(otherwise_)),
+        inputType(inputType_)
+    {}
+    
+    template <class V>
+    static ParseResult parse(const V& value, const ParsingContext& ctx) {
+        using namespace mbgl::style::conversion;
+        
+        assert(isArray(value));
+        auto length = arrayLength(value);
+        if (length < 3) {
+            return CompileError {
+                "Expected at least 2 arguments, but found only " + std::to_string(length - 1) + ".",
+                ctx.key()
+            };
+        }
+        
+        // Expect odd-length array: ["match", input, 2 * (n pairs)..., otherwise]
+        if (length % 2 != 1) {
+            return CompileError {
+                "Missing final output value for \"match\" expression.",
+                ctx.key()
+            };
+        }
+        
+        auto parsedInput = parseExpression(arrayMember(value, 1), ParsingContext(ctx, {1}, {"match"}));
+        if (parsedInput.template is<CompileError>()) {
+            return parsedInput;
+        }
+        
+        auto otherwise = parseExpression(arrayMember(value, length - 1), ParsingContext(ctx, {length - 1}, {"match"}));
```

Comment:
JS parses the non `otherwise` branches first. Should be consistent one way or the other.